taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -468,53 +471,83 @@ func (self *Contract) Run(ctx vm.CallFrame, evm *vm.EVM) ([]byte, error) {
 			fmt.Println("Unable to parse getValidators input args: ", err)
 			return nil, err
 		}
-
-		res := self.getValidators(args)
-		return method.Outputs.Pack(res.Validators, res.End)
+		return method.Outputs.Pack(self.getValidators(args))
 
 	case "getDelegations":
 		var args sol.GetDelegationsArgs
 		if err = method.Inputs.Unpack(&args, input); err != nil {
 			fmt.Println("Unable to parse getDelegations input args: ", err)
 			return nil, err
 		}
-
-		res := self.getDelegations(args)
-		return method.Outputs.Pack(res.Delegations, res.End)
+		return method.Outputs.Pack(self.getDelegations(args))
 
 	case "getUndelegations":
 		var args sol.GetUndelegationsArgs
 		if err = method.Inputs.Unpack(&args, input); err != nil {
 			fmt.Println("Unable to parse getUndelegations input args: ", err)
 			return nil, err
 		}
-
-		res := self.getUndelegations(args)
-		return method.Outputs.Pack(res.Undelegations, res.End)
+		return method.Outputs.Pack(self.getUndelegations(args))
 	}
 
 	return nil, nil
 }
 
-func (self *Contract) DistributeRewards(rewardsStats *rewards_stats.RewardsStats, feesRewards *FeesRewards) {
+func (self *Contract) DistributeRewards(author *common.Address, rewardsStats *rewards_stats.RewardsStats, feesRewards *FeesRewards) {
 	// When calling DistributeRewards, internal structures must be always initialized
 	self.lazy_init()
 
 	// Calculates number of tokens to be generated as block reward
 	blockReward := bigutil.Mul(self.amount_delegated, self.yield_percentage)
 	blockReward = bigutil.Div(blockReward, bigutil.Mul(Big100, self.blocks_per_year))
 
+	votesReward := bigutil.Big0
+	authorReward := bigutil.Big0
+	trxsReward := blockReward
+	// We need to handle case for block 1
+	if rewardsStats.TotalVotesWeight > 0 {
+		// Calculate propotion between votes and transactions
+		trxsReward = bigutil.Div(bigutil.Mul(blockReward, VotesToTrasnactionsRatio), Big100)
+		votesReward = bigutil.Sub(blockReward, trxsReward)
+
+		// In case block author included more than minimum required number of votes
+		if rewardsStats.BonusVotesWeight > 0 && self.cfg.MaxBlockAuthorReward > 0 {
+			maxBonusVotes := rewardsStats.MaxVotesWeight / 3
```

Comment:
need description for what is going on in this whole if

Reply:
hmm I still dont get this  if, especially this division and why you even need to divide something / 3 ? I think you are missing 1 info in rewardsStats(and BonusVotesWeight should not be there at all), which would allow you to skip this division(by any division you get small rounding error). If you send this data:

- MinVotesWeight // min theoretical votes weight (2/3 + 1)
- MaxVotesWeight // max theoretical votes weight
- IncludedVotesWeight // weight of votes included in block

you can calculate bonusVotesWeight = IncludedVotesWeight - MinVotesWeight

Then you dont need this divison at all because you can calculate how many % is bonusVotesWeight out of MaxVotesWeight. 

Also, pls add panics into the taraxa_evm_state_api_transition_state"
- MinVotesWeight < MaxVotesWeight
- IncludedVotesWeight >= MinVotesWeight && IncludedVotesWeight <= MaxVotesWeight

Reply:
I am also confused about validatorVoteReward ? How is word "validator" related to reward calculation for block_author based on included bonus votes ? The straightforward algo here would be:

maxBlockAuthorReward := self.cfg.MaxBlockAuthorReward * votesReward / 100
blockAuthorReward := bonusVotesWeight * maxBlockAuthorReward / maxBonusWeight

when optimized for smallest rounding error (but keep the algo from above in comments so it is easy to understand):
(bonusVotesWeight * bonusVotesWeight * maxBlockAuthorReward) / (100 * maxBonusWeight)